File: 0041_First_Missing_Positive/solution.java
package cn.huang.leetcode;
/*
41. First Missing Positive


Given an unsorted integer array, find the smallest missing positive integer.

Example 1:

Input: [1,2,0]
Output: 3

Example 2:

Input: [3,4,-1,1]
Output: 2

Example 3:

Input: [7,8,9,11,12]
Output: 1

Note:

Your algorithm should run in O(n) time and uses constant extra space.

 */


/*
https://www.youtube.com/watch?v=jfb72FfxWKU

 */
public class LC_0041_FirstMissingPositive {
    public int firstMissingPositive(int[] nums) {
        if (nums == null || nums.length == 0)
            return 1;
        for (int i = 0; i < nums.length; i++) {
            while (nums[i] > 0 && nums[i] < nums.length && nums[nums[i] - 1] != nums[i]) {
                // Put num[i] to the right position num[nuims[i]-1] (nums[i] != i + 1)
                int temp = nums[nums[i] - 1];
                nums[nums[i] - 1] = nums[i];
                nums[i] = temp;
            }
        }

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != i + 1)
                return i + 1;
        }
        return nums.length + 1;
    }
}

File: 0041_First_Missing_Positive/solution.cpp
/*
41. First Missing Positive [Hard]

https://leetcode.com/problems/first-missing-positive/
*/
#include <vector>

using namespace std;
/*
https://www.youtube.com/watch?v=jfb72FfxWKU
桶排序
 */
class Solution
{
public:
    int firstMissingPositive(vector<int> &nums)
    {
        if (nums.size() == 0)
            return 1;
        for (int i = 0; i < nums.size(); i++)
        {
            while (nums[i] > 0 && nums[i] < nums.size() && nums[nums[i] - 1] != nums[i])
            {
                // Put num[i] to the right position num[nuims[i]-1] (nums[i] != i + 1)
                int temp = nums[nums[i] - 1];
                nums[nums[i] - 1] = nums[i];
                nums[i] = temp;
            }
        }

        for (int i = 0; i < nums.size(); i++)
        {
            if (nums[i] != i + 1)
                return i + 1;
        }
        return nums.size() + 1;
    }
};